### PR TITLE
agnos updater: set decompress max_length

### PR DESCRIPTION
**Pull request by andiradulescu**
This fixes the "MemoryError: Unable to allocate output buffer." when flashing big raw images. Write-up [here](https://github.com/commaai/agnos-builder/pull/275#issuecomment-2288508865).

The following changes have been implemented.
- `noop` flashes until no more data is returned - this way any data that remains in `self.buf` after `eof` is detected gets flashed in every case - since the loop will call read until buffer is empty
- decompress `max_length` is now set to `length` ([`max_length=-1`](https://docs.python.org/3/library/lzma.html#lzma.LZMADecompressor.decompress) if not set) - so the output of decompress will never be more than the size of the return
- setting `b''` when [`needs_input=false`](https://docs.python.org/3/library/lzma.html#lzma.LZMADecompressor.needs_input) - when it has more data to decompress after the previous length
- if the last decompressed output has [`eof=True`](https://docs.python.org/3/library/lzma.html#lzma.LZMADecompressor.eof) then we reached the end of file and the process stops

Flashing time for the sparse system image is similar (a little faster with `max_length=length`).

Flashing time for non-sparse raw 10GB image is ~5min.

### CODE REVIEW

File: system/hardware/tici/agnos.py [hunk 1]
```diff
@@ -83,8 +91,11 @@ def unsparsify(f: StreamingDecompressor) -> Generator[bytes, None, None]:
 
 # noop wrapper with same API as unsparsify() for non sparse images
 def noop(f: StreamingDecompressor) -> Generator[bytes, None, None]:
-  while not f.eof:
-    yield f.read(1024 * 1024)
+  while True:
+    chunk = f.read(1024 * 1024)
+    if not chunk:
```
Comment (adeebshihadeh) [inline system/hardware/tici/agnos.py:96]: I think a `len` check is more readable here

File: system/hardware/tici/agnos.py [hunk 2]
```diff
@@ -83,8 +91,11 @@ def unsparsify(f: StreamingDecompressor) -> Generator[bytes, None, None]:
 
 # noop wrapper with same API as unsparsify() for non sparse images
 def noop(f: StreamingDecompressor) -> Generator[bytes, None, None]:
-  while not f.eof:
-    yield f.read(1024 * 1024)
+  while True:
```
Comment (adeebshihadeh) [inline system/hardware/tici/agnos.py:94]: consider a walrus: https://docs.python.org/3/whatsnew/3.8.html#new-features

File: system/hardware/tici/agnos.py [hunk 3]
```diff
@@ -29,16 +29,24 @@ def __init__(self, url: str) -> None:
...
+        try:
+          compressed = next(self.it)
+        except StopIteration:
+          self.eof = True
+          break
+      else:
+        compressed = b''
+
+      out = self.decompressor.decompress(compressed, max_length=length)
+      self.buf += out
```
Comment (adeebshihadeh) [inline system/hardware/tici/agnos.py:45]:
```suggestion
      self.buf += self.decompressor.decompress(compressed, max_length=length)
```

**Review by adeebshihadeh — APPROVED** [review]

File: system/hardware/tici/agnos.py (hunk 2 above)
Comment (adeebshihadeh) [reply system/hardware/tici/agnos.py]: FYI @sshane I think it's time to introduce the walrus

Comment (sshane) [reply system/hardware/tici/agnos.py]: Yess

Comment (andiradulescu) [reply system/hardware/tici/agnos.py]: nice! 🤩

**Comment by andiradulescu** [comment]
good cleanup and thanks for applying!

File: system/hardware/tici/agnos.py (hunk 1 above)
Comment (andiradulescu) [reply system/hardware/tici/agnos.py]: did that first and then changed my mind

**Pull request merged**